[WIP] Write a JSON Schema case study for the W3C Web of Things (#18)
json-schema-org/blog: pages/posts/w3c-wot-case-study.md

---
title: "How the W3C Web of Things brings JSON Schema to the Internet of Things"
date: 2022-08-17T06:00:00+00:00
type: Case Study
authors:
  - name: Juan Cruz Viotti
    photo: /img/avatars/jviotti.webp
    link: https://twitter.com/jviottidc

> excerpt: "Using JSON Schema at the W3C Web of Things to counter the fragmentation of the Internet of Things"

Reviewer: actually "countering the fragmentation" is something we want to avoid since it starts with a negative tone. I would propose (from our input document):

> Using JSON Schema at the W3C Web of Things to create an interoperability layer so that different IoT platforms, protocols and standards can operate together.

I hope it is not too late to fit into the small part that is under the title

Author: Makes sense! Done

language: en
---

## Challenge

The Internet of Things (IoT) is the network of physical "smart" objects that
exchange data with other devices over the Internet. While communication and
interoperability are by definition the crux of the Internet of Things, the
emergence of custom or proprietary solutions results in devices that cannot
talk to each other due to differences in data interchange mechanisms.

To integrate these disparate devices, developers must work with a growing set
of protocols, serialization formats and API specifications. This results in
repetitive, non-scalable and error-prone work that is difficult to automate.

While technologies like OpenAPI and AsyncAPI largely solve this problem in the
context of Web APIs, they fall short for describing networks of non-HTTP
multi-protocol devices and do not consider different modes of interaction based
on their meaning in the physical world.

## Solution

To solve these problems, the W3C Web of Things works on providing standardized
building blocks that make use of JSON Schema.

> JSON Schema is used for validating abstractions of physical entities called

Reviewer: abstraction of a physical entity can be understood differently by different people. Not sure of the best replacement but maybe "description of network-facing capabilities of physical entities"

Author: Yeah, I think that reads well. Done

[Thing Descriptions][thing-description], and to model and describe data sent by
Internet of Things consumers and producers in a multi-protocol manner.

> We are on-board with JSON Schema since Draft 4, and even the first draft

Reviewer: Shouldn't we use another subject than "we"? There are other places with it as well.

Author: True. Other case studies use "We" only in quotations. I'll fix it everywhere

Author: Done

together.

The work has started as a discussion in the Web of Things Community Group in
2013-2014.  It has proceed to an Interest Group in 2015 that has collected the
use cases and defined the standards to be worked on. Since 2016, the working
group is working on different standards on the Web of Things with the first
publications of the Thing Description and Architecture recommendations in 2019.

At the time of this writing, the Working Group consists of 96 participants
representing 36 organizations and the the Interest Group consists of 131
participants representing 48 organizations. 

![W3C Web of Things Examples](/blog/img/posts/2022/w3c-wot-case-study/wot-examples.webp)

## Getting Started

We invite you to leverage your existing JSON Schema knowledge to play with the
Internet of Things through the use of the W3C Web of Things specifications and
tools. You might already have a smart device next to you that does not come
with a Thing Description!

If you write your own Thing Description, you can programatically interact with
your device by using the [Node.js Web of Things
implementation](https://github.com/eclipse/thingweb.node-wot), or play with it

> through the [Node-RED](https://www.youtube.com/watch?v=oAcYbJ6P9bU)

Reviewer 2: you can also link to the node-red web of things plugin (based on node-wot): https://www.npmjs.com/package/node-red-contrib-web-of-things

continued Ege Korkan.

JSON Schema is delivering exciting features, and more importantly these
features are becoming more and more consistent across implementations.  For
example, Ege Korkan added, "vocabularies are very promising and we plan to
explore them further in conjunction with Semantic Web technologies."

## Impact

The W3C Web of Things relies on JSON-LD for linking vocabularies and ontologies
that already exist on the web, such as [SAREF][saref], [Units of
Measure][units-of-measure] and [Schema.org][schema-org]. Being based on the
JSON data model, JSON Schema fits well with JSON-LD when integrating both
technologies as part of a single specification.

Adopting JSON Schema means that the W3C Web of Things does not need to invest
effort on inventing another schema language. As a consequence of the popularity
of JSON Schema, existing tooling can be often reused instead of implementing
custom parsers, validators, code generators and UI generators for all the
programming languages that the community might be interested in.

JSON Schema has proven to be a foundational block for creating higher-level
tooling that is specific to the W3C Web of Things. As a notable example, the

> community has developed a tool that makes use of JSON Schema definitions inside

Reviewer 3: Is it worth linking to the tool?

Author: I don't have the link. @egekorkan is this public?

Reviewer: Here it is https://github.com/tum-esi/testbench

Thing Descriptions to generate matching payloads for the purpose of stress and
penetration testing.

Given the growing popularity of using JSON Schema to generate web-based forms,
W3C Web of Things specifications can be used to automatically generate
dashboards and user interfaces to interact with the objects modeled by Thing
Descriptions.

## Key Impact Results

The ultimate metric of success for a standard like the W3C Web of Things is
adoption, and community adoption is not something you can just "engineer".  Ege
Korkan commented, "because we adopt JSON Schema, developers who are familiar
with it have a smaller learning curve when learning the W3C Web of Things

> standard."

Reviewer 3: Hey @egekorkan. Here I'm ideally looking at some kind of soundbite.
Would you be able to take some sort of educated guess as to how much time it saves people getting to grips with the IoT standard?

Reviewer: I actually have some sort of numbers on this. During my PhD times, I had quite a bit of students (electrical engineering and masters level) and they had to get up to speed with WoT. A good student took 1 week full time to go through my basic tutorial at https://github.com/tum-esi/wot-tutorial . I think that 1 week is too much for many people since my tutorial goes through Linux and git basics, then some JSON Schema and Node.js before starting with WoT at all. If someone is already somewhat experienced in JSON Schema and some basic Javascript, it should be 2 days full time.

Author: Nice stuff! I tried to incorporate that example here. Let me know what you think!

with it have a smaller learning curve when learning the W3C Web of Things
standard."

"During my PhD times, my MSc engineering students had to become familiar with
the W3C Web of Things. These students typically took 1 week to go through my
[tutorial](https://github.com/tum-esi/wot-tutorial). If they knew JSON Schema,
that 1 week could become 2 days", said Ege Korkan.

## W3C Web of Things

The World Wide Web Consortium (W3C) is the standardization organization behind
most of the standards used in the Web. We are the Web of Things Working and
Interest Groups, who work on the standardization of Web of Things at W3C.

The Web of Things groups aim to create an interoperability layer so that
different Internet of Things platforms, protocols and standards can operate
together.

The work has started as a discussion in the Web of Things Community Group in
2013-2014.  It has proceeded to become an Interest Group in 2015 that has
collected the use cases and defined the standards to be worked on. Since 2016,
the working group is working on different standards on the Web of Things with
the first publications of the Thing Description and Architecture

> recommendations in 2019.

Reviewer: ```suggestion
recommendations in 2019 and more that will be published by the end of this year.
```

At the time of this writing, the Working Group consists of 96 participants
representing 36 organizations and the Interest Group consists of 131
participants representing 48 organizations. 

![W3C Web of Things Examples](/blog/img/posts/2022/w3c-wot-case-study/wot-examples.webp)

## Getting Started

We invite you to leverage your existing JSON Schema knowledge to play with the
Internet of Things through the use of the W3C Web of Things specifications and

> tools. You might already have a smart device next to you that does not come

Reviewer: ```suggestion
tools. You might already have a smart device next to you, just that it does not come
```

with a Thing Description!

If you write your own Thing Description, you can programatically interact with
your device by using the [Node.js Web of Things
implementation](https://github.com/eclipse/thingweb.node-wot), or play with it
through the
[Node-RED](https://www.npmjs.com/package/node-red-contrib-web-of-things)
integration.

To learn more, head over to the
[documentation](https://www.w3.org/WoT/documentation/), watch the [introductory
videos](https://www.w3.org/WoT/videos/), have a look at the many examples
provided as part of the [Thing Description specification](thing-description)
and try them out on the [Thing Description
Playground](http://plugfest.thingweb.io/playground/) app.

> Thank you to Ege Korkan, Web of Things Researcher at TU Munich, and the W3C Web

Reviewer: ```suggestion
Thank you to Ege Korkan, Web of Things Researcher at Siemens AG, Munich, the W3C Web
```

> of Things, for allowing and enabling us to share this case study with you.

Reviewer: ```suggestion
of Things Working and Interest Groups and Coralie Mercier from the W3C Marketing and Communications Team for allowing and enabling us to share this case study with you.
```

To solve these problems, the W3C Web of Things works on providing standardized
building blocks that make use of JSON Schema.

JSON Schema is used for validating descriptions of network-facing capabilities
of physical entities called [Thing Descriptions][thing-description], and to
model and describe data sent by Internet of Things consumers and producers in a
multi-protocol manner.

The W3C Web of Things specifications are on board with JSON Schema since Draft
4, and even the first draft versions already catered to data models of Internet
of Things devices.  "Currently, we are on JSON Schema Draft 7 and expect to
move to JSON Schema
[2020-12](https://json-schema.org/specification-links.html#understanding-draft-names-and-numbers)
or a newer one when starting our new charter in February 2023," said Ege
Korkan, Research Scientist at Siemens and W3C Specifications editor.

"We observe JSON Schema becoming more and more stable over the years and
Postman's support to the community gives us even more confidence on it,"
continued Ege Korkan.

JSON Schema is delivering exciting features, and more importantly these
features are becoming more and more consistent across implementations.  For

> example, Ege Korkan added, "vocabularies are very promising and we plan to

Reviewer: ```suggestion
example, Ege Korkan added, "JSON Schema vocabularies are very promising and we plan to
```

better to be specific since vocabulary is a very commonly used term in WoT and Semantic Web. If you prefer another way to say that, I would be also fine :)

Adopting JSON Schema means that the W3C Web of Things does not need to invest
effort on inventing another schema language. As a consequence of the popularity
of JSON Schema, existing tooling can be often reused instead of implementing
custom parsers, validators, code generators and UI generators for all the
programming languages that the community might be interested in.

JSON Schema has proven to be a foundational block for creating higher-level
tooling that is specific to the W3C Web of Things. As a notable example, the
community has developed a tool called
[testbench](https://github.com/tum-esi/testbench) that makes use of JSON Schema
definitions inside Thing Descriptions to generate matching payloads for the
purpose of stress and penetration testing.

Given the growing popularity of using JSON Schema to generate web-based forms,
W3C Web of Things specifications can be used to automatically generate
dashboards and user interfaces to interact with the objects modeled by Thing
Descriptions.

## Key Impact Results

The ultimate metric of success for a standard like the W3C Web of Things is
adoption, and community adoption is not something you can just "engineer".  Ege

> Korkan commented, "because we adopt JSON Schema, developers who are familiar

Reviewer: ```suggestion
Korkan commented, "since we adopt JSON Schema, developers who are familiar
```

sounds better to me when this sentence is considered on its own

with it have a smaller learning curve when learning the W3C Web of Things
standard."

"During my PhD times, my MSc engineering students had to become familiar with
the W3C Web of Things. These students typically took 1 week to go through my
[tutorial](https://github.com/tum-esi/wot-tutorial). If they knew JSON Schema,

> that 1 week could become 2 days", said Ege Korkan.

Reviewer: ```suggestion
that 1 week would become 2 days", said Ege Korkan.
```

## W3C Web of Things

> The World Wide Web Consortium (W3C) is the standardization organization behind
> most of the standards used in the Web. We are the Web of Things Working and
> Interest Groups, who work on the standardization of Web of Things at W3C.

Reviewer 3: ```suggestion
The World Wide Web Consortium (W3C) is the standardization organization behind
most of the standards used in the Web. The W3C Web of Things Working and
Interest Groups work on the standardization of Web of Things at W3C.
```

Missed one

The Web of Things groups aim to create an interoperability layer so that
different Internet of Things platforms, protocols and standards can operate
together.

> The work has started as a discussion in the Web of Things Community Group in
> 2013-2014.  It has proceeded to become an Interest Group in 2015 that has
> collected the use cases and defined the standards to be worked on. Since 2016,
> the working group is working on different standards on the Web of Things with
> the first publications of the Thing Description and Architecture
> recommendations in 2019 and more that will be published by the end of this
> year.

Reviewer 3: ```suggestion
The work started as a discussion in the Web of Things Community Group in
2013-2014.  It proceeded to become an Interest Group in 2015 that
collected the use cases and defined the standards to be worked on. Since 2016,
the working group is working on different standards on the Web of Things with
the first publications of the Thing Description and Architecture
recommendations in 2019 and more that will be published by the end of this
year.
```

Minor gramatical corrections.

At the time of this writing, the Working Group consists of 96 participants
representing 36 organizations and the Interest Group consists of 131
participants representing 48 organizations. 

![W3C Web of Things Examples](/blog/img/posts/2022/w3c-wot-case-study/wot-examples.webp)

## Getting Started

> We invite you to leverage your existing JSON Schema knowledge to play with the

Reviewer 3: ```suggestion
The W3C Web of Things Working Group invite you to leverage your existing JSON Schema knowledge to play with the
```

Reviewer: ```suggestion
The W3C Web of Things Working Group invites you to leverage your existing JSON Schema knowledge to play with the
```

Author: Done!

model and describe data sent by Internet of Things consumers and producers in a
multi-protocol manner.

> The W3C Web of Things are on-board with JSON Schema since Draft 4, and even the

Reviewer 4: s/are on-board/specifications are on board/

> first draft versions were quite enough to model data models of Internet of

Reviewer 4: s/were quite enough to model/already catered to/

> Things devices.  "Currently, we are on Draft 7 and expect to move to 2020-12 or

Reviewer 4: What does "2020-12" mean? How does it relate to draft numbers which appear earlier in the sentence? I'm afraid it may be meaningless otherwise to people outside of the group.

Reviewer 5: It might be good to include a link to https://json-schema.org/specification-links.html (which has a "Understanding draft names and numbers" section, because it's a surprisingly complex topic).

Author: Make sense. Done!

Reviewer 5: @koalie you asked how to resolve a comment- you should be able to see a "Resolve conversation" button below this comment reply.

Reviewer 4: No I don't :( Can anyone else do that instead of me? I only see "review changes" at the same level as "conversation", "commits", "checks"; and no button at all below the comment reply.

> a newer one when starting our new charter in February 2023." - Ege Korkan

Reviewer 4: s/." - Ege Korkan/," said Ege Korkan, Research Scientist at Siemens and W3C Specifications editor.

"We observe JSON Schema becoming more and more stable over the years and

> Postman's support to the community gives us even more confidence on it." - Ege

Reviewer 4: s/on it." - Ege/in it," continued Ege/

> Korkan 

Reviewer 4: s/Korkan/Korkan./

JSON Schema is delivering exciting features, and more importantly these
features are becoming more and more consistent across implementations.  For
example, "vocabularies are very promising and we plan to explore them further
in conjunction with Semantic Web technologies." - Ege Korkan

> 

Reviewer 4: I suggest you remove "- Ege Korkan", and instead write: For example, Ege Korkan added, "vocab[...]"

The ultimate metric of success for a standard like the W3C Web of Things is
adoption, and community adoption is not something you can just "engineer".
"Because we adopt JSON Schema, developers who are familiar with it have a
smaller learning curve when learning the W3C Web of Things standard." - Ege
Korkan

> 

Reviewer 4: I suggest this rewrite:
Ege Korkan commented, "Because we use JSON Schema, developers who are familiar with it have a lesser learning curve when adopting the W3C Web of Things standard."

## W3C Web of Things

> World Wide Web Consortium (W3C) is the standardization organization behind most

Reviewer 4: s/World/The World/

of the standards used in the Web. We are the Web of Things Working and Interest

> Groups, who work on the standardization of Web of Things at the W3C.

Reviewer 4: s/at the W3C/at W3C/

The Web of Things groups aim to create an interoperability layer so that
different Internet of Things platforms, protocols and standards can operate
together.

The work has started as a discussion in the Web of Things Community Group in

> 2013-2014.  It has proceed to an Interest Group in 2015 that has collected the

Reviewer 4: s/proceed/proceeded to become/

use cases and defined the standards to be worked on. Since 2016, the working
group is working on different standards on the Web of Things with the first
publications of the Thing Description and Architecture recommendations in 2019.

At the time of this writing, the Working Group consists of 96 participants

> representing 36 organizations and the the Interest Group consists of 131

Reviewer 4: s/the the/the/
